File: Sources/PolyReadLinguistics/Chunker.swift
import Foundation
import PolyReadCore

/// §6.2 — "Chunking runs AFTER phonemization. v1 had this backwards. `tokens` is
/// capped at `[1, 3…512]` — 510 phonemes plus two boundary zeros — and you cannot
/// know a paragraph's phoneme length until you phonemize it."
public struct Chunker: Sendable {

    /// §7.1 — 510 phonemes, the two boundary zeros excluded.
    public static let budget = 510
    /// §6.2 — "back off to the most recent sentence end within the last 25% of
    /// the budget."
    public static let sentenceBackoffFraction = 0.25

    let vocabulary: KokoroVocabulary

    public init(vocabulary: KokoroVocabulary = .shared) {
        self.vocabulary = vocabulary
    }

    public struct Result: Sendable {
        public let chunks: [PhonemizedChunk]
        /// Symbols the phonemizer emitted that the vocabulary does not know. Not
        /// fatal — they are dropped — but a long list means the phoneme set and
        /// the vocabulary disagree, which is §0.3 territory.
        public let unknownSymbols: [String: Int]
    }

    /// Phonemizes and chunks one block.
    public func chunk(block: Block, using phonemizer: any Phonemizer, tagger: POSTagger = POSTagger()) throws -> Result {
        guard block.role.isSpoken else { return Result(chunks: [], unknownSymbols: [:]) }

        let tokens = SpanInvariant.tokens(of: block.spokenText).map(String.init)
        guard !tokens.isEmpty else { return Result(chunks: [], unknownSymbols: [:]) }

        let posTags = phonemizer.capabilities.resolvesHomographs ? [] : tagger.tag(tokens: tokens)
        let words = try phonemizer.phonemize(tokens: tokens, posTags: posTags)

        guard words.count == tokens.count else {
            // §0.3 — "If it only returns a flat phoneme string, stop and raise it."
            // A phonemizer that cannot keep one entry per word cannot support
            // word-level highlighting, and guessing an alignment would produce a
            // highlight that drifts instead of an error that says why.
            throw PolyReadError.phonemizerLacksWordGrouping
        }

        // Encode once. `wordPhonemeRanges` indexes into this.
        var tokenIDs: [Int32] = []
        var ranges: [Range<Int>] = []
        var unknown: [String: Int] = [:]
        // Word separator. Kokoro reads the space as a word boundary; its own
        // duration lands in the gap between two words, which is exactly where
        // Timeline expects a gap to be.
        let space = vocabulary.symbolToID[" "]

        for (index, word) in words.enumerated() {
            if index > 0, let space { tokenIDs.append(space) }
            let start = tokenIDs.count
            let (encoded, missing) = vocabulary.encode(word.phonemes)
            for symbol in missing { unknown[symbol, default: 0] += 1 }
            tokenIDs.append(contentsOf: encoded)
            // A word that encoded to nothing still needs a range, or the 1:1
            // alignment with `Block.spans` breaks. An empty range at the right
            // place gives it zero duration and keeps every later index correct.
            ranges.append(start..<tokenIDs.count)
        }

        let chunks = split(
            blockID: block.id,
            tokenIDs: tokenIDs,
            ranges: ranges,
            tokens: tokens
        )
        return Result(chunks: chunks, unknownSymbols: unknown)
    }

    /// The split itself. Separated out so it is testable without a phonemizer.
    func split(
        blockID: UUID,
        tokenIDs: [Int32],
        ranges: [Range<Int>],
        tokens: [String]
    ) -> [PhonemizedChunk] {
        guard !ranges.isEmpty else { return [] }
        let total = tokenIDs.count
        if total <= Self.budget {
            return [
                PhonemizedChunk(
                    blockID: blockID,
                    tokens: tokenIDs,
                    wordPhonemeRanges: rebased(ranges, by: ranges[0].lowerBound),
                    spanOffset: 0
                )
            ]
        }

        // §6.2 — "Emit chunks as close to uniform length as the text allows — the
        // style vector is selected by phoneme count (§7.1), so wildly varying
        // chunk lengths give wandering prosody." So aim for total/n rather than
        // filling each chunk to 510 and leaving a 40-phoneme runt at the end.
        let chunkCount = Int((Double(total) / Double(Self.budget)).rounded(.up))
        let target = min(Self.budget, Int((Double(total) / Double(chunkCount)).rounded(.up)))
        let backoff = Int(Double(target) * Self.sentenceBackoffFraction)

        var chunks: [PhonemizedChunk] = []
        var wordIndex = 0

        while wordIndex < ranges.count {
            let chunkStart = ranges[wordIndex].lowerBound
            var end = wordIndex
            var lastSentenceEnd: Int?

            while end < ranges.count {
                let wouldBe = ranges[end].upperBound - chunkStart
                // Always take at least one word, or a pathological single word
                // longer than the budget would loop forever.
                if end > wordIndex, wouldBe > target { break }
                if end > wordIndex, wouldBe > Self.budget { break }
                if isSentenceEnd(tokens[end]) { lastSentenceEnd = end }
                end += 1
            }

            // §6.2 — back off to the most recent sentence end, but only if it is
            // inside the last 25% of the budget. Backing off further would make
            // this chunk much shorter than its neighbours, which is the thing
            // uniformity is trying to avoid.
            var cut = end
            if let sentenceEnd = lastSentenceEnd, sentenceEnd + 1 < end {
                let lengthAtSentenceEnd = ranges[sentenceEnd].upperBound - chunkStart
                if lengthAtSentenceEnd >= target - backoff {
                    cut = sentenceEnd + 1
                }
            }
            cut = max(cut, wordIndex + 1)

            let sliceEnd = ranges[cut - 1].upperBound
            chunks.append(
                PhonemizedChunk(
                    blockID: blockID,
                    tokens: Array(tokenIDs[chunkStart..<sliceEnd]),
                    wordPhonemeRanges: rebased(Array(ranges[wordIndex..<cut]), by: chunkStart),
                    spanOffset: wordIndex
                )
            )
            wordIndex = cut
        }

        return chunks
    }

    private func rebased(_ ranges: [Range<Int>], by offset: Int) -> [Range<Int>] {
        ranges.map { ($0.lowerBound - offset)..<($0.upperBound - offset) }
    }

    /// Sentence-final on the *source* token, not the phonemes — "Putnam." is a
    /// sentence end and "et al" is not, and only the orthography knows that.
    func isSentenceEnd(_ token: String) -> Bool {
        let trimmed = token.trimmingCharacters(in: CharacterSet(charactersIn: "\"'”’)]}"))
        guard let last = trimmed.last else { return false }
        guard last == "." || last == "!" || last == "?" else { return false }
        // An initial ("J. S. Mill") or a surviving abbreviation is not a sentence.
        let core = String(trimmed.dropLast())
        if core.count <= 1 { return false }
        if core.allSatisfy(\.isUppercase) && core.count <= 3 { return false }
        return true
    }
}

/// Runs the whole of Agent B over a document: §5 then §6.
public struct LinguisticsPipeline: Sendable {
    let normalizer = Normalizer()
    let chunker: Chunker
    let phonemizer: any Phonemizer

    public init(phonemizer: any Phonemizer = PhonemizerFactory.make(), vocabulary: KokoroVocabulary = .shared) {
        self.phonemizer = phonemizer
        self.chunker = Chunker(vocabulary: vocabulary)
    }

    public struct Output: Sendable {
        public let blocks: [Block]
        public let reflow: ReflowDocument
        /// Main stream, in document order — body, headings, captions.
        public let mainChunks: [PhonemizedChunk]
        /// §4.5 — footnote bodies are chunked and prosody-run too, but kept out
        /// of the main stream so a page of notes does not land mid-sentence.
        public let footnoteChunks: [UUID: [PhonemizedChunk]]
        public let unknownSymbols: [String: Int]
    }

    public func run(blocks rawBlocks: [Block]) throws -> Output {
        let normalized = normalizer.normalize(rawBlocks)
        // Reflow ranges have to exist before Phase A mints WordTimings from these
        // spans, so the layout happens here rather than in the UI.
        let (reflow, blocks) = ReflowDocumentBuilder.build(blocks: normalized)

        var mainChunks: [PhonemizedChunk] = []
        var footnoteChunks: [UUID: [PhonemizedChunk]] = [:]
        var unknown: [String: Int] = [:]

        for block in blocks {
            guard block.role.isSpoken else { continue }
            let result = try chunker.chunk(block: block, using: phonemizer)
            for (symbol, count) in result.unknownSymbols { unknown[symbol, default: 0] += count }

            if block.role == .footnoteBody {
                footnoteChunks[block.id] = result.chunks
            } else {
                mainChunks.append(contentsOf: result.chunks)
            }
        }

        #if DEBUG
        for chunk in mainChunks + footnoteChunks.values.flatMap({ $0 }) {
            assert(chunk.tokens.count <= Chunker.budget, "chunk over budget: \(chunk.tokens.count)")
        }
        #endif

        return Output(
            blocks: blocks,
            reflow: reflow,
            mainChunks: mainChunks,
            footnoteChunks: footnoteChunks,
            unknownSymbols: unknown
        )
    }
}

File: web/src/linguistics/chunker.ts
import { PolyReadError } from "../core/errors";
import { tokensOf } from "../core/spanInvariant";
import { isSpoken, newID, type Block, type PhonemizedChunk, type TokenRange } from "../core/types";
import { tagTokens } from "./posTagger";
import type { Phonemizer } from "./phonemizer";
import { defaultVocabulary, type KokoroVocabulary } from "./vocabulary";

/**
 * §6.2 — "Chunking runs AFTER phonemization. v1 had this backwards. `tokens` is
 * capped at `[1, 3…512]` — 510 phonemes plus two boundary zeros — and you
 * cannot know a paragraph's phoneme length until you phonemize it."
 */
export const BUDGET = 510;

/**
 * Where a chunk stops taking sentences, in phoneme tokens — about fifteen
 * seconds of speech.
 *
 * 510 is the model's ceiling, not a good operating point. Kokoro's prosody is
 * at its best on a sentence or a few of them and audibly flattens and rushes as
 * an input approaches the cap, which is why the reference clients all feed it
 * sentence by sentence rather than filling the window. A chunk this size also
 * renders in a fraction of the time a full one does, which is what the reader
 * is waiting on at the start of a document and after every seek into audio
 * that does not exist yet.
 */
export const TARGET = 250;

/**
 * A chunk shorter than this takes the next sentence even past `TARGET`, so a
 * three-word sentence is not rendered as an utterance of its own.
 */
export const MIN_CHUNK = 60;

/** A single sentence longer than this is divided at its clauses. */
export const LONG_SENTENCE = 340;

export interface ChunkResult {
  chunks: PhonemizedChunk[];
  /**
   * Symbols the phonemizer emitted that the vocabulary does not know. Not fatal
   * — they are dropped — but a long list means the phoneme set and the
   * vocabulary disagree, which is §0.3 territory.
   */
  unknownSymbols: Record<string, number>;
}

/**
 * Abbreviations that survive §5 — its table covers the ones that need *reading*
 * differently, which is a different list from the ones that merely end in a
 * period.
 */
const ABBREVIATIONS = new Set([
  "vol", "vols", "no", "nos", "ed", "eds", "ch", "chap", "chaps", "fig",
  "figs", "tab", "tabs", "pp", "esp", "vs", "mr", "mrs", "ms", "dr",
  "prof", "st", "jr", "sr", "rev", "trans", "repr", "sec", "secs", "nn",
  "cit", "ff", "inc", "dept", "univ", "co", "corp", "et", "al", "cf",
]);

/**
 * Sentence-final on the *source* token, not the phonemes — "Putnam." is a
 * sentence end and "Vol." is not, and only the orthography knows that.
 *
 * Being wrong here is cheap in one direction and not the other: a missed
 * sentence end just means the split falls on a word boundary instead, while a
 * false one puts a chunk break inside a sentence. So the abbreviation tests are
 * deliberately generous.
 */
export function isSentenceEnd(token: string): boolean {
  const trimmed = token.replace(/["'”’)\]}]+$/u, "");
  const last = trimmed[trimmed.length - 1];
  if (last !== "." && last !== "!" && last !== "?") return false;
  if (last !== ".") return true;

  const core = trimmed.slice(0, -1);
  // An initial: "J. S. Mill".
  if (core.length <= 1) return false;
  // An internal period: "U.S.", and anything §5 did not rewrite.
  if (core.includes(".")) return false;
  return !ABBREVIATIONS.has(core.toLowerCase());
}

export class Chunker {
  private readonly vocabulary: KokoroVocabulary;

  constructor(vocabulary: KokoroVocabulary = defaultVocabulary) {
    this.vocabulary = vocabulary;
  }

  /** Phonemizes and chunks one block. */
  async chunk(block: Block, phonemizer: Phonemizer): Promise<ChunkResult> {
    if (!isSpoken(block.role)) return { chunks: [], unknownSymbols: {} };

    const tokens = tokensOf(block.spokenText);
    if (tokens.length === 0) return { chunks: [], unknownSymbols: {} };

    const posTags = phonemizer.capabilities.resolvesHomographs ? [] : tagTokens(tokens);
    const words = await phonemizer.phonemize(tokens, posTags);

    if (words.length !== tokens.length) {
      // §0.3 — "If it only returns a flat phoneme string, stop and raise it."
      // A phonemizer that cannot keep one entry per word cannot support
      // word-level highlighting, and guessing an alignment would produce a
      // highlight that drifts instead of an error that says why.
      throw new PolyReadError("phonemizerLacksWordGrouping");
    }

    // Encode once. `wordPhonemeRanges` indexes into this.
    const tokenIDs: number[] = [];
    const ranges: TokenRange[] = [];
    const joined: boolean[] = [];
    const unknown: Record<string, number> = {};
    // Word separator. Kokoro reads the space as a word boundary; its own
    // duration lands in the gap between two words, which is exactly where
    // Timeline expects a gap to be.
    const space = this.vocabulary.spaceID;

    for (let index = 0; index < words.length; index++) {
      const word = words[index];
      const { tokens: encoded, unknown: missing } = this.vocabulary.encode(word.phonemes);
      for (const symbol of missing) unknown[symbol] = (unknown[symbol] ?? 0) + 1;
      // No space before a word eSpeak welded to the previous one — the stream
      // stays exactly as eSpeak produced it — and none for a word with nothing
      // to say, which would otherwise leave two spaces in a row.
      const separated = !word.joined && encoded.length > 0 && tokenIDs.length > 0;
      if (separated && space !== undefined) tokenIDs.push(space);
      const start = tokenIDs.length;
      for (const id of encoded) tokenIDs.push(id);
      // A word that encoded to nothing still needs a range, or the 1:1
      // alignment with `Block.spans` breaks. An empty range at the right place
      // gives it zero duration and keeps every later index correct.
      ranges.push({ start, end: tokenIDs.length });
      joined.push(word.joined === true);
    }

    const chunks = splitChunks(block.id, tokenIDs, ranges, tokens, joined);
    return { chunks, unknownSymbols: unknown };
  }
}

function rebased(ranges: readonly TokenRange[], offset: number): TokenRange[] {
  return ranges.map((r) => ({ start: r.start - offset, end: r.end - offset }));
}

const CLAUSE_END = /[,;:—–]["'”’)\]]*$/u;

/**
 * The split itself. Separated out so it is testable without a phonemizer.
 *
 * A chunk is whole sentences, packed until the next one would take it past
 * `TARGET`. Kokoro renders each chunk as one utterance with its own intonation
 * contour, so a cut anywhere but a sentence end is heard: the voice drops as if
 * the sentence were over, pauses, and starts the rest on a fresh breath. Only a
 * sentence too long to render well is divided, at a clause if it has one and at
 * a word boundary if it does not — and never between two words eSpeak welded
 * together (`joined`), which share one group of phonemes.
 */
export function splitChunks(
  blockID: string,
  tokenIDs: readonly number[],
  ranges: readonly TokenRange[],
  tokens: readonly string[],
  joined: readonly boolean[] = [],
): PhonemizedChunk[] {
  if (ranges.length === 0) return [];
  const count = ranges.length;
  /** A chunk may begin at word `w`. */
  const canStart = (w: number): boolean => w <= 0 || w >= count || !joined[w];
  /** Phoneme tokens in words `[from, to)`. */
  const length = (from: number, to: number): number => ranges[to - 1].end - ranges[from].start;

  // Sentences.
  const sentences: Array<[number, number]> = [];
  let from = 0;
  for (let w = 0; w < count; w++) {
    if (w === count - 1 || (isSentenceEnd(tokens[w] ?? "") && canStart(w + 1))) {
      sentences.push([from, w + 1]);
      from = w + 1;
    }
  }

  // Sentences too long to render well, divided at clauses, then at words.
  const pieces: Array<[number, number]> = [];
  for (const [a, b] of sentences) {
    if (length(a, b) <= LONG_SENTENCE) {
      pieces.push([a, b]);
      continue;
    }
    for (const [c, d] of pack(a, b, (w) => CLAUSE_END.test(tokens[w] ?? ""), TARGET)) {
      if (length(c, d) <= LONG_SENTENCE) {
        pieces.push([c, d]);
        continue;
      }
      // No clause to cut at: as even as the words allow.
      const parts = Math.ceil(length(c, d) / TARGET);
      pieces.push(...pack(c, d, () => true, Math.ceil(length(c, d) / parts)));
    }
  }

  // Whole pieces, packed up to TARGET.
  const spans: Array<[number, number]> = [];
  for (const piece of pieces) {
    const current = spans[spans.length - 1];
    if (current) {
      const combined = length(current[0], piece[1]);
      if (combined <= TARGET || (length(current[0], current[1]) < MIN_CHUNK && combined <= BUDGET)) {
        current[1] = piece[1];
        continue;
      }
    }
    spans.push([piece[0], piece[1]]);
  }
  // A short last sentence rides with the one before it rather than alone.
  if (spans.length >= 2) {
    const last = spans[spans.length - 1];
    const previous = spans[spans.length - 2];
    if (length(last[0], last[1]) < MIN_CHUNK && length(previous[0], last[1]) <= BUDGET) {
      previous[1] = last[1];
      spans.pop();
    }
  }

  return spans.map(([a, b]) => makeChunk(blockID, tokenIDs, ranges, a, b));

  /**
   * Words `[a, b)` in runs of at most `target` tokens, cutting after a word
   * `isCut` accepts, or before the word that would pass `BUDGET` when none
   * does.
   */
  function pack(a: number, b: number, isCut: (w: number) => boolean, target: number): Array<[number, number]> {
    const out: Array<[number, number]> = [];
    let start = a;
    while (start < b) {
      let end = b;
      let lastCut = -1;
      for (let w = start; w < b; w++) {
        const size = length(start, w + 1);
        if (w > start && size > BUDGET) {
          end = lastCut > start ? lastCut : previousStart(w, start);
          break;
        }
        if (size > target && lastCut > start) {
          end = lastCut;
          break;
        }
        if (w + 1 < b && canStart(w + 1) && isCut(w)) lastCut = w + 1;
      }
      out.push([start, end]);
      start = end;
    }
    return out;
  }

  /** The last word at or before `w` a chunk may start at, but never `start` itself. */
  function previousStart(w: number, start: number): number {
    let at = w;
    while (at > start + 1 && !canStart(at)) at--;
    return Math.max(at, start + 1);
  }
}

function makeChunk(
  blockID: string,
  tokenIDs: readonly number[],
  ranges: readonly TokenRange[],
  from: number,
  to: number,
): PhonemizedChunk {
  const chunkStart = ranges[from].start;
  let sliceEnd = ranges[to - 1].end;
  let chunkRanges = rebased(ranges.slice(from, to), chunkStart);

  // A single "word" longer than the whole budget is not English — it is a URL,
  // or two pages of a scan that came back without spaces. The model's input is
  // capped at [1, 3…512], so it has to be cut somewhere; cutting is better
  // than throwing away the paragraph around it.
  if (sliceEnd - chunkStart > BUDGET) {
    sliceEnd = chunkStart + BUDGET;
    chunkRanges = chunkRanges.map((range) => {
      const start = Math.min(range.start, BUDGET);
      return { start, end: Math.max(start, Math.min(range.end, BUDGET)) };
    });
  }

  return {
    id: newID(),
    blockID,
    tokens: tokenIDs.slice(chunkStart, sliceEnd),
    wordPhonemeRanges: chunkRanges,
    spanOffset: from,
  };
}
